### PR TITLE
New testsuite for testing Kernel Live Patching in SLE Micro

### PR DESCRIPTION
**Pull request by pablo-herranz**
- Related ticket: https://progress.opensuse.org/issues/98234
- Needles: 
- Verification run: https://openqa.suse.de/t9725261

### CODE REVIEW

File: tests/kernel/qa_test_klp.pm [hunk 1]
```diff
@@ -26,7 +27,11 @@ sub run {
     }
 
     if (script_run('[ -d /lib/modules/$(uname -r)/build ]') != 0) {
-        zypper_call('in -l kernel-devel');
+        if (is_transactional) {
+            trup_call('pkg in -l kernel-devel');
```
Comment (mdoucha) [inline tests/kernel/qa_test_klp.pm:31]: It'd be better to add a helper function like `install_package()` that'll do this if/else internally.

File: tests/kernel/qa_test_klp.pm [hunk 2]
```diff
@@ -36,8 +41,11 @@ sub run {
     (is_sle(">12-sp1") || !is_sle) ? $self->select_serial_terminal() : select_console('root-console');
 
     add_suseconnect_product("sle-sdk") if (is_sle('<12-SP5'));
-    zypper_call('in -l autoconf automake gcc git make');
-
+    if (is_transactional) {
+        trup_call('in -l autoconf automake gcc git make');
```
Comment (mdoucha) [inline tests/kernel/qa_test_klp.pm:45]: Shouldn't the argument start with `pkg in`?

**Review by mdoucha — CHANGES_REQUESTED** [review]

File: tests/kernel/qa_test_klp.pm (hunk 2 above)
Comment (pablo-herranz) [reply tests/kernel/qa_test_klp.pm:45]: You're right, I'll fix it

File: tests/kernel/qa_test_klp.pm (hunk 1 above)
Comment (pablo-herranz) [reply tests/kernel/qa_test_klp.pm:31]: Yes, sounds like a good idea!

**Comment by mdoucha** [comment]
The `install_package()` function will be useful for other tests as well so it'd be better to put it somewhere in `lib/`. Preferably in a separate commit.

File: tests/kernel/qa_test_klp.pm [hunk 3]
```diff
@@ -15,7 +15,17 @@ use base 'opensusebasetest';
...
+use transactional 'trup_call';
+
+sub install_package {
+    my $command = $_[0];
+    if (is_transactional) {
+        trup_call('pkg in -l ' . $command);
+    } else {
+        zypper_call('in -l ' . $command);
+    }
+}
```
Comment (foursixnine) [inline tests/kernel/qa_test_klp.pm:28]: as @mdoucha is suggesting, please move this to `lib/Wrappers/Packages.pm` (or something that isn't just dumping it on `lib/`

File: tests/microos/libzypp_config.pm [hunk 4]
```diff
@@ -11,15 +11,19 @@ use base "opensusebasetest";
...
     shift->select_serial_terminal();
     unless (check_var('FLAVOR', 'JeOS-for-AArch64') || check_var('FLAVOR', 'JeOS-for-RPi')) {
         assert_script_run 'egrep -x "^solver.onlyRequires ?= ?true" /etc/zypp/zypp.conf';
         assert_script_run 'egrep -x "^rpm.install.excludedocs ?= ?yes" /etc/zypp/zypp.conf';
     }
-    assert_script_run sprintf('egrep -x "^multiversion ?=%s" /etc/zypp/zypp.conf', is_jeos ? ' ?provides:multiversion\(kernel\)' : '');
+    if (is_jeos || is_transactional) {
+        assert_script_run "echo 'multiversion = provides:multiversion(kernel)' >> /etc/zypp/zypp.conf";
+        assert_script_run "echo 'multiversion.kernels = latest' >> /etc/zypp/zypp.conf";
+        assert_script_run "echo \"LIVEPATCH_KERNEL='always'\" >> /etc/sysconfig/livepatching";
```
Comment (foursixnine) [inline tests/microos/libzypp_config.pm:25]: Maybe:

```suggestion
        assert_script_run 'echo LIVEPATCH_KERNEL="always" >> /etc/sysconfig/livepatching';
```

**Review by foursixnine — CHANGES_REQUESTED** [review]

File: tests/microos/libzypp_config.pm (hunk 4 above)
Comment (pablo-herranz) [reply tests/microos/libzypp_config.pm]: Changed :)

File: tests/kernel/qa_test_klp.pm (hunk 3 above)
Comment (pablo-herranz) [reply tests/kernel/qa_test_klp.pm]: I've made the changes, let's see if this works this way :crossed_fingers:

**Comment by pdostal** [comment]
> The `install_package()` function will be useful for other tests as well so it'd be better to put it somewhere in `lib/`. Preferably in a separate commit.

@mdoucha maybe even separate pull request?

**Comment by mdoucha** [comment]
> > The `install_package()` function will be useful for other tests as well so it'd be better to put it somewhere in `lib/`. Preferably in a separate commit.
> 
> @mdoucha maybe even separate pull request?

If the separate PR gets merged faster than this whole thing, then sure, why not.

**Review by czerw — APPROVED** [review]

**Comment by jlausuch** [comment]
This is a bit tricky, you can't just run `qa_test_klp.pm` as it won't find the needed packages `autoconf automake gcc git make. I would suggest to close this one and let kernel qa team take over as they might have better ideas how to handle live patching on sle micro.

**Pull request closed without merging**